src/model.rs

```diff
@@ -724,6 +724,138 @@ impl Collections {
+    ///                                                                                    
+    ///  VJ:1 (Mon-Sun)   09:00-10:00       10:00-11:00                                    
+    ///  VJ:2 (Mon-Fri)                                       12:00-13:00       14:00-15:00
+    ///  VJ:3 (Sat-Sun)                                       12:30-13:30       14:30-15:30
+    ///
+    /// Example 4 might be a valid use case of prolongation trip but is not
+    /// handled in the current implementation since the validity patterns are
+    /// different; this is undefined behavior (most likely, some prolongation
+    /// trip will be forbidden by the default rule).
+    pub fn enhance_pickup_dropoff(&mut self) {
+        let mut vj_idxs: Vec<Idx<VehicleJourney>> = self
+            .vehicle_journeys
+            .iter()
+            // Filtering out vehicle journeys without any stop time allowing us
+            // below to call stop_times[0] and stop_times[stop_times.len() - 1]
+            // safely.
+            .filter(|(_, vj)| !vj.stop_times.is_empty())
+            .map(|(idx, _)| idx)
+            .collect();
+        // Init all vehicle journeys with:
+        // - no drop-off on first stop
+        // - no pick-up on last stop
+        for vj_idx in &vj_idxs {
+            let mut vehicle_journey = self.vehicle_journeys.index_mut(*vj_idx);
+            let first_stop = &mut vehicle_journey.stop_times[0];
+            first_stop.drop_off_type = 1;
+            let stop_times_len = vehicle_journey.stop_times.len();
+            let last_stop = &mut vehicle_journey.stop_times[stop_times_len - 1];
+            last_stop.pickup_type = 1;
+        }
+        // Keep only vehicle journeys with a `block_id` (concerned with
+        // prolongation trip).
+        vj_idxs.retain(|vj_idx| self.vehicle_journeys[*vj_idx].block_id.is_some());
+        // Sort the vehicule journeys by block_id, and if equal, order them by
+        // departure time.
+        vj_idxs.sort_by(|vj1_idx, vj2_idx| {
+            let vj1 = &self.vehicle_journeys[*vj1_idx];
+            let vj2 = &self.vehicle_journeys[*vj2_idx];
+            match vj1.block_id.cmp(&vj2.block_id) {
+                Ordering::Equal => vj1.stop_times[0]
+                    .departure_time
+                    .cmp(&vj2.stop_times[0].departure_time),
+                ordering => ordering,
+            }
+        });
+        for (prev_vj_idx, next_vj_idx) in vj_idxs.iter().zip(vj_idxs.iter().skip(1)) {
+            let prev_vj = &self.vehicle_journeys[*prev_vj_idx];
+            let next_vj = &self.vehicle_journeys[*next_vj_idx];
+            if prev_vj.block_id != next_vj.block_id {
+                // we can discard two vehicle journeys which are not part of the
+                // same prolongation trip.
+                continue;
+            }
+            let stop_times_len = prev_vj.stop_times.len();
+            let last_stop = &prev_vj.stop_times[stop_times_len - 1];
```

Comment:
you can probably use (deref to slice, and you know it's not empty)
```suggestion
            let last_stop = &prev_vj.stop_times.last().unwrap();
```

Reply:
First of all, the use of `.unwrap()` here is not riskier than using slices `[]` since slice's operator panics if out-of-bound. So what you propose is equivalent and I do indeed prefer it.

I've actually made the modification everywhere to replace with `.first()`/`.last()` (`.first_mut()`/`.last_mut()` when it applies). See 2f90dfbf71468da64af9fd87cc05f538ac17cd2e.